File: subly_android/app/src/main/java/net/cynreub/subly/ui/payment/addedit/AddEditPaymentMethodViewModel.kt
package net.cynreub.subly.ui.payment.addedit

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.launch
import net.cynreub.subly.domain.model.PaymentMethod
import net.cynreub.subly.domain.model.PaymentType
import net.cynreub.subly.domain.repository.PaymentMethodRepository
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class AddEditPaymentMethodViewModel @Inject constructor(
    private val paymentMethodRepository: PaymentMethodRepository,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val paymentMethodId: String? = savedStateHandle["paymentMethodId"]

    private val _uiState = MutableStateFlow(AddEditPaymentMethodUiState())
    val uiState: StateFlow<AddEditPaymentMethodUiState> = _uiState.asStateFlow()

    init {
        if (paymentMethodId != null) {
            loadPaymentMethod(paymentMethodId)
        }
    }

    private fun loadPaymentMethod(id: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            try {
                val uuid = UUID.fromString(id)
                val paymentMethod = paymentMethodRepository.getPaymentMethodById(uuid)
                    .catch { e ->
                        _uiState.value = AddEditPaymentMethodUiState(
                            isLoading = false,
                            error = "Failed to load payment method: ${e.message}"
                        )
                    }
                    .firstOrNull()

                if (paymentMethod != null) {
                    _uiState.value = AddEditPaymentMethodUiState(
                        nickname = paymentMethod.nickname,
                        selectedType = paymentMethod.type,
                        lastFourDigits = paymentMethod.lastFourDigits ?: "",
                        isLoading = false,
                        isEditMode = true,
                        paymentMethodId = paymentMethod.id
                    )
                } else {
                    _uiState.value = AddEditPaymentMethodUiState(
                        isLoading = false,
                        error = "Payment method not found"
                    )
                }
            } catch (e: IllegalArgumentException) {
                _uiState.value = AddEditPaymentMethodUiState(
                    isLoading = false,
                    error = "Invalid payment method ID"
                )
            }
        }
    }

    fun onNicknameChange(nickname: String) {
        _uiState.value = _uiState.value.copy(
            nickname = nickname,
            nicknameError = null
        )
    }

    fun onLastFourDigitsChange(digits: String) {
        // Only allow digits and limit to 4 characters
        val filteredDigits = digits.filter { it.isDigit() }.take(4)
        _uiState.value = _uiState.value.copy(
            lastFourDigits = filteredDigits,
            lastFourDigitsError = null
        )
    }

    fun showTypeDialog() {
        _uiState.value = _uiState.value.copy(showTypeDialog = true)
    }

    fun dismissTypeDialog() {
        _uiState.value = _uiState.value.copy(showTypeDialog = false)
    }

    fun onTypeSelected(type: PaymentType) {
        _uiState.value = _uiState.value.copy(
            selectedType = type,
            showTypeDialog = false
        )
    }

    fun onSaveClick() {
        if (!validateForm()) {
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true)

            try {
                val paymentMethod = PaymentMethod(
                    id = _uiState.value.paymentMethodId ?: UUID.randomUUID(),
                    nickname = _uiState.value.nickname.trim(),
                    type = _uiState.value.selectedType,
                    lastFourDigits = _uiState.value.lastFourDigits.takeIf { it.isNotBlank() },
                    icon = null // Icon not used in current implementation
                )

                if (_uiState.value.isEditMode) {
                    paymentMethodRepository.updatePaymentMethod(paymentMethod)
                } else {
                    paymentMethodRepository.insertPaymentMethod(paymentMethod)
                }

                _uiState.value = _uiState.value.copy(
                    isSaving = false,
                    saveSuccess = true
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isSaving = false,
                    error = "Failed to save payment method: ${e.message}"
                )
            }
        }
    }

    private fun validateForm(): Boolean {
        val nickname = _uiState.value.nickname.trim()
        val lastFourDigits = _uiState.value.lastFourDigits

        var isValid = true

        // Validate nickname
        if (nickname.isBlank()) {
            _uiState.value = _uiState.value.copy(nicknameError = "Nickname is required")
            isValid = false
        }

        // Validate last four digits (optional, but if provided must be 4 digits)
        if (lastFourDigits.isNotBlank() && lastFourDigits.length != 4) {
            _uiState.value = _uiState.value.copy(lastFourDigitsError = "Must be exactly 4 digits")
            isValid = false
        }

        return isValid
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }
}

File: subly_ios/Subly/Subly/UI/PaymentMethods/AddEdit/AddEditPaymentMethodViewModel.swift
import Foundation
import Combine
import Observation

@Observable
final class AddEditPaymentMethodViewModel {
    var uiState = AddEditPaymentMethodUiState()

    private let paymentMethodRepository: PaymentMethodRepository
    private var cancellables = Set<AnyCancellable>()

    init(paymentMethodRepository: PaymentMethodRepository, paymentMethodId: UUID? = nil) {
        self.paymentMethodRepository = paymentMethodRepository
        uiState.paymentMethodId = paymentMethodId
        uiState.isEditMode = paymentMethodId != nil
        if let id = paymentMethodId { loadPaymentMethod(id: id) }
    }

    private func loadPaymentMethod(id: UUID) {
        uiState.isLoading = true
        paymentMethodRepository.getPaymentMethodById(id)
            .receive(on: DispatchQueue.main)
            .first()
            .sink { [weak self] method in
                guard let self, let method else {
                    self?.uiState.isLoading = false
                    self?.uiState.error = "Payment method not found"
                    return
                }
                uiState.nickname = method.nickname
                uiState.selectedType = method.type
                uiState.lastFourDigits = method.lastFourDigits ?? ""
                uiState.isLoading = false
            }
            .store(in: &cancellables)
    }

    func onNicknameChange(_ value: String) {
        uiState.nickname = value
        uiState.nicknameError = nil
    }

    func onTypeSelected(_ type: PaymentType) {
        uiState.selectedType = type
        uiState.showTypePicker = false
    }

    func onLastFourChange(_ value: String) {
        let filtered = value.filter { $0.isNumber }
        uiState.lastFourDigits = String(filtered.prefix(4))
        uiState.lastFourDigitsError = nil
    }

    func onSaveClick(onSuccess: @escaping () -> Void) {
        uiState.nicknameError = nil
        uiState.lastFourDigitsError = nil
        uiState.error = nil

        var hasErrors = false
        if uiState.nickname.trimmingCharacters(in: .whitespaces).isEmpty {
            uiState.nicknameError = "Nickname is required"
            hasErrors = true
        }
        let digits = uiState.lastFourDigits
        if !digits.isEmpty && digits.count != 4 {
            uiState.lastFourDigitsError = "Must be exactly 4 digits"
            hasErrors = true
        }
        if hasErrors { return }

        let method = PaymentMethod(
            id: uiState.paymentMethodId ?? UUID(),
            nickname: uiState.nickname.trimmingCharacters(in: .whitespaces),
            type: uiState.selectedType,
            lastFourDigits: uiState.lastFourDigits.isEmpty ? nil : uiState.lastFourDigits
        )

        uiState.isSaving = true
        Task { @MainActor in
            do {
                if uiState.isEditMode {
                    try await paymentMethodRepository.updatePaymentMethod(method)
                } else {
                    try await paymentMethodRepository.insertPaymentMethod(method)
                }
                uiState.isSaving = false
                onSuccess()
            } catch {
                uiState.isSaving = false
                uiState.error = "Failed to save: \(error.localizedDescription)"
            }
        }
    }
}
